02-peer-protocol.md

```diff
@@ -50,10 +67,374 @@ pubkey corresponding to the funding output nothing prevents duplicative channel
 ids.
 
 
+### `channel_id`, v2
+
+For channels established using the v2 protocol, the `channel_id` is the
+SHA256(lesser-revocation-basepoint || greater-revocation-basepoint),
+where the lesser and greater is based off the order of the basepoint.
+The basepoints are compact DER-encoded public keys.
+
+If the peer's revocation basepoint is unknown (e.g. `open_channel2`
+and `accept_channel2`), a temporary `channel_id` should be found by
+using a zeroed out basepoint for the unknown peer.
+
+#### Rationale
+These values must be remembered by both peers for correct operation anyway.
+They're known from the first exchange of messages, obviating the need for
+a `temporary_channel_id`. Finally, by mixing information from both sides,
+they avoid `channel_id` collisions.
+
+
+## Interactive Transaction Construction
+
+Interactive transaction construction allows two peers to collaboratively
+build a transaction for broadcast.  This protocol is the foundation
+for dual-funded channels establishment (v2).
+
+### Set-Up and Vocabulary
+
+There are two parties to a transaction construction: an *initiator*
+and a *non-initiator*.
+ The *initiator* is the peer which initiates the protocol, e.g.
+for channel establishment v2 the *initiator* would be the peer which
+sends `open_channel2`; for a close, the *initiator* sends `shutdown`.
+
+The protocol makes the following assumptions:
+
+- The `feerate` for the transaction is known.
+- The `dust_limit` for the transaction is known.
+- The `nLocktime` for the transaction has been negotiated, or is established
+  via convention.
+- The transaction version is 2.
+
+### Fee Responsibility
+
+The *initiator* is responsible for paying the fees for the following fields,
+to be referred to as the `common fields`.
+
+  - version
+  - segwit marker + flag
+  - input count
+  - output count
+  - locktime
+
+The rest of the transaction bytes' fees are the responsibility of
+the peer who contributed that input or output via `tx_add_input` or
+`tx_add_output`, at the agreed upon `feerate`.
+
+
+### Overview
+
+The *initiator* initiates the interactive transaction construction
+protocol with `tx_add_input`. The *non-initiator* responds with any
+of `tx_add_input`, `tx_add_output`, `tx_rm_input`, `tx_rm_output`, or
+ `tx_complete`. The protocol continues with the synchronous exchange
+of interactive transaction protocol messages until both nodes have sent
+and received a consecutive `tx_complete`.
+
+Once peers have exchanged consecutive `tx_complete`s, the
+interactive transaction construction protocol is considered concluded.
+Both peers should construct the transaction and fail the negotiation
+if an error is discovered.
```

Comment:
Ah. So 'fail the negotiation' does have an explicit meaning for the implementation that I've got in c-lightning. My bad for not explicitly including it here!

A 'negotiation fail' consists of two things:
- first, sending a 'warning' message with the error
- next, disconnecting with the peer.

The disconnect has the benefit of resetting the state such that the peer can reconnect and try again from scratch.

Reply:
Thanks, that's more clear.

I still think that a dedicated message (`abort_tx`?) would make sense though as:

- we may want to send non-fatal warnings (without disconnecting) during the interactive-tx sessions for other reasons
- it could be useful to abort without necessarily disconnecting (one of the parties changed their mind)
- it's more future-proof since it gives us a message that we can easily extend in the future to add data that could help automate retries

WDYT?

Reply:
That's because you're looking at it through the lens of c-lightning which has a single channel per peer ;)
Imagine I already have a channel with you and payments are steadily flowing through it.
In parallel, we start a dual funding flow.
But actually I've run out of utxos (because I'm also splicing in with someone else in parallel) so I want to abort.
I don't want to have to disconnect to tell you that, because that disrupts our other channel that is processing payments.
Does that make sense?

Reply:
Ah yes, you present a very persuasive argument here @t-bast! c-lightning will indeed have this same problem as of our next release lol.

Ok, cool I'll reconfigure it to add an `abort_tx` call, which should be equivalent to hanging up (also).

Reply:
Actually, question. Does the `error` message not serve the same purpose here? It should result in the same desired behavior signal: this negotiation is fatally flawed and cannot progress forward.

Reply:
Or a `warning`? Either of these is useful for signaling "this conversation is over, let's start again".

Reply:
I don't think we should use `warning` for that, there could be reasons to send informative `warning`s without failing the negotiation. Maybe `error` would work though, but I still think it's slightly cleaner to have an explicit `abort`, as aborting isn't necessarily an error case (and we may want in the future to enrich the `tx_abort` message with specific options / fields).

I'm not sure about that yet, but I'll be able to report back on this once I've made progress on the implementation!